### PR TITLE
[partitioner] Fix must_be_in_backward corner cases

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132755
* #132638
* __->__ #134002
* #134003

Preparation PR for https://github.com/pytorch/pytorch/pull/132638

"must_be_in_backward" fails the partitioner, if partitioner picks this node as saved_values.

The fix is to prevent partitioner to pick those nodes during nodes classification.

It's hard to make a test without making effectful ops in backward "must_be_in_backward", which will be testing this ( https://github.com/pytorch/pytorch/pull/132638 )

### CODE REVIEW

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -251,6 +249,14 @@ def _is_backward_state(node: fx.Node) -> bool:
     return node.op == "placeholder" and isinstance(node.meta.get("val"), BackwardState)
 
 
+def _has_tag_must_be_in_backward(node: fx.Node) -> bool:
+    return node.meta.get("partitioner_tag", None) == "must_be_in_backward"
+
+
+def _must_be_in_backward(node: fx.Node) -> bool:
+    return _has_tag_must_be_in_backward(node)
```
Comment (bdhirsh) [inline torch/_functorch/partitioners.py:257]: `_has_tag_must_be_in_backward` seems unnecessary? (we can just inline it into `_must_be_in_backward`)

**Review by bdhirsh — APPROVED** [review]
seems ok to me

File: torch/_functorch/partitioners.py (hunk 1 above)
Comment (IvanKobzarev) [reply torch/_functorch/partitioners.py:257]: My idea here is that in future _must_be_in_backward will accumulate multiple conditions, one of them will be `_has_tag_must_be_in_backward`

**Pull request closed without merging**